go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,360 @@
+}
+
+func newIndexType(dbTable dbObjType, contType controllerType, ownerObjectType ownerType) IndexType {
+	value, ok := dbIndexes[dbTable][contType][ownerObjectType]
+	if !ok {
+		panic("requested index is unknown")
+	}
+	return IndexType{dbTable, contType, ownerObjectType, value}
+}
+
+func (it IndexType) GetIDs() []string {
+	return it.ids
+}
+
+func (it IndexType) Equivalent(it2 IndexType) bool {
+	return it.ownerObjectType == it2.ownerObjectType && it.dbTable == it2.dbTable && it.contType == it2.contType
+}
+
+// IndexTypes should only be created here
+
+var AddressSetEgressFirewallDNS = newIndexType(addressSet, defaultController, EgressFirewallDNSOwnerType)
+
+var AddressSetHybridNodeRoute = newIndexType(addressSet, defaultController, HybridNodeRouteOwnerType)
+
+var AddressSetEgressQoS = newIndexType(addressSet, defaultController, EgressQoSOwnerType)
+
+var AddressSetNetworkPolicy = newIndexType(addressSet, defaultController, NetworkPolicyOwnerType)
+
+var AddressSetNamespace = newIndexType(addressSet, defaultController, NamespaceOwnerType)
+
+var AddressSetEgressIP = newIndexType(addressSet, defaultController, EgressIPOwnerType)
+
+var AddressSetEgressService = newIndexType(addressSet, defaultController, EgressServiceOwnerType)
+
+// DbIndex is a structure representing a set of db object ExternalIDs, used to uniquely identify
+// an object in the db (as a primary index), and to filter out a subset of objects based on some IDs
+// (as a secondary index).
+// DbIndex consists of 3 parts:
+// - IndexT defines which IDs are used for given object, and references a list of ID keys from dbIndexes map.
+// IndexT.ownerObjectType will be written to ExternalIDs[OwnerObjectType]
+// - OwnerControllerName define who manages given object. It is required to communicate with the other
+// controllers that create/update/delete db objects.
+// - ObjectIDs uniquely identify objects within the same OwnerControllerName and IndexT.
+//
+//  dbIndex := NewDbIndex(AddressSetEgressFirewallDNS, "DefaultController",
+//		map[string]string{
+//			SecondaryIndexKey: "dns.name",
+// 			AddressSetIPFamilyID: "ipv4"
+//	})
+// uses AddressSetEgressFirewallDNS = newIndexType(addressSet, defaultController, EgressFirewallDNSOwnerType),
+// which in turn uses the following IDs based on dbIndexes map
+//  dbIndexes[addressSet][defaultController][EgressFirewallDNSOwnerType] = {
+//		// dnsName
+//		SecondaryIndexKey,
+//		AddressSetIPFamilyID,
+//	}
+//
+// dbIndex will be mapped to the following ExternalIDs
+// {
+//		"k8s.ovn.org/controller": "DefaultController"
+//		"k8s.ovn.org/object-type": "EgressFirewallDNS" (value of EgressFirewallDNSOwnerType)
+//		"k8s.ovn.org/name": "dns.name"
+// 		"k8s.ovn.org/ipFamily": "ipv4"
+//		"k8s.ovn.org/id": "DefaultController:EgressFirewallDNS:dns.name:ipv4"
+// }
+//
+// To ensure correct usage of DbIndex, and field verification, always use NewDbIndex or GetDbIndexFromExternalIDs
+// to create the object.
+type DbIndex struct {
```

Comment:
nit: I did like ObjectIds as a more descriptive to what this really is rather than DbIndex